@xen-orchestra/xapi/vm.js

```diff
@@ -514,12 +514,27 @@ class Vm {
     }
   }
 
-  async snapshot($defer, vmRef, { cancelToken = CancelToken.none, ignoreNobakVdis = false, name_label } = {}) {
+  async snapshot(
+    $defer,
+    vmRef,
+    { cancelToken = CancelToken.none, ignoreNobakVdis = false, name_label, unplugVusbs = false } = {}
+  ) {
     const vm = await this.getRecord('VM', vmRef)
 
+    const isHalted = vm.power_state === 'Halted'
+
+    // requires the VM to be halted because it's not possible to re-plug VUSB on a live VM
+    if (unplugVusbs && isHalted) {
+      await asyncMap(vm.VUSBs, async ref => {
+        const vusb = await this.getRecord('VUSB', ref)
+        await vusb.$call('destroy')
+        $defer.call(this, 'call', 'VUSB.create', vusb.VM, vusb.USB_group, vusb.other_config)
```

Comment:
Is the console.error message if the VUSB.create fails enough ? I think it should fails the backup or at least be presented to the user

Reply:
It should not fail the backup because the snapshot as succeeded.

This function can be run in other contexts that backups and has no easy way to report the error to the calling code, what do you propose?

Reply:
A snapshot has been created, but the VM is now in a broken state, as usb devices are often mandatory security device. I think it can't be ignored, so even if an error seems a little too serious I think it's better than hiding it

Reply:
I have a few ideas regarding this:
- a dedicated `defer` in `@xen-orchestra/xapi` with custom `onError`
- hierarchical logging support in `@xen-orchestra/log`, let's discuss this in the next meeting.

Reply:
Yes, but we cannot fail the backup for this.

Reply:
This problem can also happen for other stuff like ignored VDIs.

I think this PR is good enough as it is and we'll do what I said before to solve this issue for all deferred errors :slightly_smiling_face:

Reply:
fine by me